doc/developer/design/20230607_shift_status_updates.md

```diff
@@ -0,0 +1,151 @@
+
+When dropping a replica, the storage controller will need to retrieve all source/sink dataflows that have been observed to be associated with the given storage instance. While this state is currently maintained in `mz_storage_client::controller::rehydration::RehydrationTask`, we can lift it up into `RehydrationStorageClient` in order for the storage controller to access it without duplicating the data. Upon dropping an instance, the storage controller can be sure that it won't receive any more `StorageResponse`'s from that instance. As a result, late status updates after an instance is dropped are guaranteed not to occur.
+
+Immediately after the coordinator announces initialization is complete, the storage controller will be updated to look for any ingestion/export dataflows that belong to an instance with 0 replicas and set their status to `paused`. This should allow `environmentd` to correctly set the status for sources/sinks in scenarios where it crashes before it was able to write the `paused` status to persist.
+
+In order to support the future in which storage clusters may have multiple replicas, we will need to update the schemas for the `mz_{source|sink}_status_history` relations to include a `replica_id`. This is necessary for the reason that dropping a replica in a storage cluster with multiple replicas shouldn't set the status of a source, for example, to `paused`. Isolating status updates for a source/sink to a specific replica allows any derived relations such as `mz_{source|sink}_statuses` to make determinations based on all the replicas in the cluster.
+
+### Rollout strategy
+
+There is opportunity for the above work to be broken down into a couple of smaller chunks. The PRs involved may look like the following:
+
+1. Shift status writes to the storage controller
+2. Update `mz_{source|sink}_status_history` and any derived relations to account for a `replica_id`, which is necessary for multi-replica storage clusters
+
+## Alternatives
+
+<!--
+Similar to the Description section. List of alternative approaches considered,
+pros/cons or why they were not chosen.
+-->
+
+1. Add a graceful shutdown mechanism for storage workers
+    - Rejected due to the fact that graceful shutdown is not guaranteed to occur
+
+## Open questions
+
+<!--
+Anything currently unanswered that needs specific focus. This section may be
+expanded during the doc meeting as other unknowns are pointed out. These
+questions may be technical, product, or anything in-between.
+-->
+
+- Should we control the rate of `StatusUpdates` from the worker or batch writes at the controller?
```

Comment:
The former feels like the essential one to me, though we'd probably benefit from doing both if it's not too difficult.